Index: src/app/org/openmeetings/app/remote/LanguageService.java
===================================================================
--- src/app/org/openmeetings/app/remote/LanguageService.java	(revision 4707)
+++ src/app/org/openmeetings/app/remote/LanguageService.java	(working copy)
@@ -50,6 +50,10 @@
 		return fieldLanguageDaoImpl.getLanguages();
 	}
 
+    public FieldLanguage getLanguage(Long language_id) {
+		return fieldLanguageDaoImpl.getFieldLanguageById(language_id);
+	}
+
 	/**
 	 * get all fields of a given Language_id
 	 * 
Index: WebContent/src/modules/chat/overAllChatTabBottom.lzx
===================================================================
--- WebContent/src/modules/chat/overAllChatTabBottom.lzx	(revision 4707)
+++ WebContent/src/modules/chat/overAllChatTabBottom.lzx	(working copy)
@@ -13,7 +13,7 @@
             this._ctext.setAttribute('text',txt);
         </method>
         <customInputtext name="_ctext" width="${ parent.width-2 }" height="$once{ parent.height-2 }"
-            x="1" y="1" bgcolor="white" >
+            x="1" y="1" bgcolor="white" isrtl="${parent.parent._inverseWording.getValue()}">
             <handler name="onkeyup" args="key">
                 <![CDATA[
                 // 13 is return
@@ -62,7 +62,7 @@
                 parent.objMessage[2] = 'newtextmessage';
                 parent.objMessage[3] = hib.userobject.login;
                 var message = this.parent.parent.message.getText();
-                if (parent.parent._inverseWording.getValue()) message = canvas.reverseWordingsBySplit(message)
+                //if (parent.parent._inverseWording.getValue()) message = canvas.reverseWordingsBySplit(message)
                 parent.objMessage[4] = message;
                 parent.objMessage[5] = canvas.currentusercolor;
                 parent.objMessage[6] = canvas.publicSID;
Index: WebContent/src/modules/conference/whiteboard/base/baseDrawWords.lzx
===================================================================
--- WebContent/src/modules/conference/whiteboard/base/baseDrawWords.lzx	(revision 4707)
+++ WebContent/src/modules/conference/whiteboard/base/baseDrawWords.lzx	(working copy)
@@ -2,7 +2,7 @@
 <library>
 
 <class name="styleableInput" extends="customInputtext" x="1" y="1" bgcolor="0xFFFFFF" multiline="true"
-		width="${ parent.width-2 }" height="${ parent.height-2 }" >
+		width="${ parent.width-2 }" height="${ parent.height-2 }">
 
 	<attribute name="startAdjustHight" value="false" type="boolean" />
 	
@@ -78,7 +78,7 @@
 			
         	this.refObj.drawTextField(this._textview._textview.getURLEncodedText(),
                     this._textview._textview.width,
-                    this._textview._textview.height,x,y);
+                    this._textview._textview.height,x,y,this._inverseWording.getValue());
                     
         } else if (this._textview._textview.getURLEncodedText().length!=0 && this.inittext.length!=0){
             
@@ -97,7 +97,7 @@
         	this.refObj.setTextFieldByName(this.initTextName,
                 this._textview._textview.getURLEncodedText(),
                 this._textview._textview.width,
-                x,y,this._textview._textview.height);
+                x,y,this._textview._textview.height, this._inverseWording.getValue());
                 
         } else {
 			if ($debug) Debug.write("text smaller then 0",this._textview._textview.getURLEncodedText());
@@ -119,11 +119,14 @@
 		//Events would set the height to 
 		//text field default height
 		this.refObj.letterObjectClose(this);
+        if ($debug) Debug.write("oninit this.refObj.isrtl ",this.refObj.isrtl);
+
         var t = new lz.styleableInput(this._textview,{
 	                name:'_textview',text:this.inittext,
 	                fontstyle:this.refObj.currentlayerstyle,
 	                fontsize:this.refObj.currentlayersize,
-	                fgcolor:this.refObj.currentlayerletter
+	                fgcolor:this.refObj.currentlayerletter,
+                    isrtl:this.refObj.isrtl
                 });
         //if ( this.inittext.length==0 ) this.setAttribute('height',40);
 		if ( this.inittext.length!=0 ) {
@@ -146,12 +149,14 @@
         //if ($debug) Debug.write("callNew height-4 ",this.getHeight());
         
 		this._textview._textview.destroy();
+        if ($debug) Debug.write("callNew this.refObj.isrtl ",this.refObj.isrtl);
 		var t = new lz.styleableInput(this._textview,{
 			        setFocusOnInit:false,
                     name:'_textview',text:txxt,
                     fontstyle:this.refObj.currentlayerstyle,
                     fontsize:this.refObj.currentlayersize,
-                    fgcolor:this.refObj.currentlayerletter
+                    fgcolor:this.refObj.currentlayerletter,
+                    isrtl:this.refObj.isrtl
                 });
         this.setAttribute("height",h);
 	</method>
@@ -165,6 +170,21 @@
     </method>
     
     <view name="_textview" width="${ parent.width }" height="${ parent.height }" />
+
+    <labelCheckbox name="_inverseWording" x="2" y="${ parent.height - this.height}" labelid="580">
+        <attribute name="doUpdate" value="false" type="boolean" />
+        <handler name="oninit">
+            this.setValue(parent.refObj.isrtl);
+            this.setAttribute('doUpdate', true);
+        </handler>
+        <handler name="onvalue" args="v">
+            if (this.doUpdate) {
+                parent.refObj.setAttribute('isrtl', this.getValue());
+                parent._textview._textview.setAttribute('isrtl', this.getValue());
+            }
+        </handler>
+        <labelTooltip labelid="581" />
+    </labelCheckbox>
     
     <resizestate name="rs" >
     	<text x="2" y="2" text="${ 'w: '+parent.width+' h: '+parent.height }" bgcolor="white"/>
Index: WebContent/src/modules/conference/whiteboard/base/baseDrawLetter.lzx
===================================================================
--- WebContent/src/modules/conference/whiteboard/base/baseDrawLetter.lzx	(revision 4707)
+++ WebContent/src/modules/conference/whiteboard/base/baseDrawLetter.lzx	(working copy)
@@ -48,6 +48,8 @@
   	
   	<!-- "plain", "bold" , "italic" or "bolditalic". -->
   	<attribute name="currentlayerstyle" value="plain" type="string" setter="this.setNewLayerStyle(currentlayerstyle)" />
+
+    <attribute name="isrtl" value="$once{canvas.isrtl}" type="boolean"/>
     
     <method name="setNewLayerLetter" args="cl">
         this.currentlayerletter=cl;
@@ -110,22 +112,22 @@
 		]]>
 	</method>
   	
-		<method name="drawTextField" args="textforfield,width,height,x,y" >
+		<method name="drawTextField" args="textforfield,width,height,x,y,isrtl" >
 
 			//Debug.write("drawTextField: ",textforfield,width,x,y);
 			var tempO = this.newTextField(this, 'letter' + this.getCounter(),
 					textforfield, width, height, x, y, this.currentlayersize,
-					this.currentlayerletter, this.currentlayerstyle);
+					this.currentlayerletter, this.currentlayerstyle,isrtl);
 			this.layers.push(tempO);
 			this.drawTextFieldregister(textforfield, width, x, y,
 					this.currentlayersize, this.currentlayerletter,
 					this.currentlayerstyle, tempO.height, this
-							.getSWFDocumentStatus(), this.getZIndex());
+							.getSWFDocumentStatus(), this.getZIndex(), isrtl);
 		</method>
 	
 	<!-- editModus -->
-	<method name="setTextFieldByName" args="txtName,textforfield,width,x,y,height" >
-		if ($debug) Debug.write("drawTextFieldByName ",txtName,textforfield,width,x,y);
+	<method name="setTextFieldByName" args="txtName,textforfield,width,x,y,height,isrtl" >
+		if ($debug) Debug.write("drawTextFieldByName ",txtName,textforfield,width,x,y,isrtl);
 		//this.checkForChanges(txtName,textforfield,width,x,y,height);
 		this.edittextLetterObject.setAttribute('text',textforfield);
 		this.edittextLetterObject.setTextEncoded(textforfield);
@@ -135,6 +137,7 @@
 		this.edittextLetterObject.setAttribute('fontsize',this.currentlayersize);
 		this.edittextLetterObject.setAttribute('fgcolor',this.currentlayerletter);
 		this.edittextLetterObject.setAttribute('fontstyle',this.currentlayerstyle);
+        this.edittextLetterObject.setAttribute('textalign', isrtl ? 'right' : 'left');
 		this.updateObjectBounds();
 		this.updateByEditLetterObject();
 		//this.setModus('hand');
@@ -212,7 +215,8 @@
 			    this.baseactionobjectList[eg][1] = this.edittextLetterObject.text;
 			    this.baseactionobjectList[eg][2] = this.edittextLetterObject.fgcolor;
 			    this.baseactionobjectList[eg][3] = this.edittextLetterObject.fontsize;
-			    this.baseactionobjectList[eg][4] = this.edittextLetterObject.fontstyle;			
+			    this.baseactionobjectList[eg][4] = this.edittextLetterObject.fontstyle;
+			    this.baseactionobjectList[eg][5] = this.edittextLetterObject.isrtl;
 				this.baseactionobjectList[eg][this.baseactionobjectList[eg].length-5] = this.edittextLetterObject.x
 				this.baseactionobjectList[eg][this.baseactionobjectList[eg].length-4] = this.edittextLetterObject.y;
 				this.baseactionobjectList[eg][this.baseactionobjectList[eg].length-3] = this.edittextLetterObject.width;
@@ -237,6 +241,7 @@
 		editObj.setAttribute('fontsize',actionObject[3]);
 		editObj.setAttribute('fgcolor',actionObject[2]);
 		editObj.setAttribute('fontstyle',actionObject[4]);
+        editObj.setAttribute('isrtl',actionObject[5]);
 	</method>
 	
 	<method name="checkForChanges" args="txtName,textforfield,width,x,y,height">
@@ -252,12 +257,13 @@
 		this.newTextField(obj,actionObject[actionObject.length-1],actionObject[1],
 		              actionObject[actionObject.length-3],actionObject[actionObject.length-2],
 		              actionObject[actionObject.length-5],actionObject[actionObject.length-4],
-		              actionObject[3],actionObject[2],actionObject[4]);
+		              actionObject[3],actionObject[2],actionObject[4], actionObject[5]);
 		              
 		this.doSWFDocumentStatus(this.currentlayer,actionObject[actionObject.length-7]);
 	</method>
 	
-		<method name="newTextField" args="obj,naming,textforfield,width,height,x,y,fontsize,fgcolor,fontstyle">
+		<method name="newTextField" args="obj,naming,textforfield,width,height,x,y,fontsize,fgcolor,fontstyle,isrtl">
+            if ($debug) Debug.write(obj,naming,textforfield,width,height,x,y,fontsize,fgcolor,fontstyle,isrtl);
 			this.currentlayer = new lz.whiteBoardTextField(obj, {
 						name : naming,
 						x : x,
@@ -268,29 +274,31 @@
 						text : textforfield,
 						fontsize : fontsize,
 						fgcolor : fgcolor,
-						fontstyle : fontstyle
+						fontstyle : fontstyle,
+                        textalign: isrtl ? 'right' : 'left'
 					});
 			this.currentlayer.setTextEncoded(textforfield);
 			//Add Layer to global Layer  	
 			return this.currentlayer;
 		</method>
 	
-	<method name="drawTextFieldregister" args="textforfield,width,x,y,fontsize,fgcolor,fontstyle,height,swfObj,zIndex">
+	<method name="drawTextFieldregister" args="textforfield,width,x,y,fontsize,fgcolor,fontstyle,height,swfObj,zIndex,isrtl">
 		var actionObject = new Array();
 	    actionObject[0] = "letter";
 	    actionObject[1] = textforfield;
 	    actionObject[2] = fgcolor;
 	    actionObject[3] = fontsize;
 	    actionObject[4] = fontstyle;
-	    actionObject[5] = zIndex;//-8
-	    actionObject[6] = swfObj;//-7
-		actionObject[7] = this.counter;//-6
-	    actionObject[8] = x;//-5
-	    actionObject[9] = y;//-4
-	    actionObject[10] = width;//-3
-	    actionObject[11] = height;//-2
-	    actionObject[12] = this.currentlayer.name;//-1	    	    
-	    this.baseactionobjectList.push(actionObject);	
+        actionObject[5] = isrtl;
+	    actionObject[6] = zIndex;//-8
+	    actionObject[7] = swfObj;//-7
+		actionObject[8] = this.counter;//-6
+	    actionObject[9] = x;//-5
+	    actionObject[10] = y;//-4
+	    actionObject[11] = width;//-3
+	    actionObject[12] = height;//-2
+	    actionObject[13] = this.currentlayer.name;//-1
+	    this.baseactionobjectList.push(actionObject);
 	    this.checkStepLayers();
 	    this.onsharedMessage('draw',actionObject);	
 	</method>
Index: WebContent/src/modules/conference/tabcontent/chat/chatTabItemEmoticons.lzx
===================================================================
--- WebContent/src/modules/conference/tabcontent/chat/chatTabItemEmoticons.lzx	(revision 4707)
+++ WebContent/src/modules/conference/tabcontent/chat/chatTabItemEmoticons.lzx	(working copy)
@@ -35,12 +35,12 @@
 	        //Debug.write("downloadurl "+downloadurl);
 	        this._userpic.setAttribute('src',downloadurl);	
 	        
-	        if(canvas.isrtl == 'true'){
+	        if(canvas.isrtl){
 	         	if($debug) Debug.write("chatTabItemEmoticons RTL is set");
 	       
-	        	this.header_user.setAttribute('textalign', 'right');
-	        	this.header_time.setAttribute('align', 'left');
-	        	this.header_time.setAttribute('x', '40');
+	        	//this.header_user.setAttribute('textalign', 'right');
+	        	//this.header_time.setAttribute('align', 'left');
+	        	//this.header_time.setAttribute('x', '40');
 	        	this.multi.setAttribute('textalign', 'right');
 	        	this.multi.setAttribute('align', 'right');
 	        }
@@ -63,7 +63,7 @@
 	</image>
 	
 	
-	<text text="$once{ parent.username }" x="40" resize="false" fontsize="10"
+	<text text="$once{ canvas.isrtl ? canvas.reverseWordingsBySplit(parent.username) : parent.username }" x="40" resize="false" fontsize="10"
 		  width="$once{ parent.width - 40 }" bgcolor="$once{ parent.usercolor }"
 		  name="header_user" />
 		  
@@ -149,7 +149,7 @@
 	<!-- space between each line of text -->
 	<attribute name="leading" value="5" type="number" />
 	
-	<text multiline="true" x="40" y="20" width="$once{ parent.width - 40 }" fontsize="11" selectable="true" name="multi" >
+	<text multiline="true" x="40" y="20" width="$once{ parent.width - 40 }" fontsize="11" selectable="true" name="multi">
 		<handler name="oninit">
 			<![CDATA[
 				
Index: WebContent/src/modules/conference/tabcontent/chat/chatTabBottom.lzx
===================================================================
--- WebContent/src/modules/conference/tabcontent/chat/chatTabBottom.lzx	(revision 4707)
+++ WebContent/src/modules/conference/tabcontent/chat/chatTabBottom.lzx	(working copy)
@@ -13,15 +13,9 @@
 				this._ctext.setAttribute('text',txt);
 			</method>
 			<customInputtext name="_ctext" width="${ parent.width-2 }" height="$once{ parent.height-2 }"
-				x="1" y="1" bgcolor="white" >
+				x="1" y="1" bgcolor="white" isrtl="${parent.parent._inverseWording.getValue()}">
 				<handler name="onkeyup" args="key">
 	                <![CDATA[
-	                if ($debug) Debug.write("onkeyup");
-	                
-	                var reverse = canvas.reverseAll(parent.getText());
-	                if ($debug) Debug.write("reverse : ", reverse);
-	                
-	               // parent.setText(reverse);
 	                // 13 is return
 	                if ((key==27) || (key==13)) {
 	                	this.parent.parent._button._button.onclick.sendEvent();
Index: WebContent/src/base/mainMethods.lzx
===================================================================
--- WebContent/src/base/mainMethods.lzx	(revision 4707)
+++ WebContent/src/base/mainMethods.lzx	(working copy)
@@ -9,7 +9,7 @@
             
             if($debug) Debug.write("reInit");
             
-            if(canvas.isrtl == 'true'){
+            if(canvas.isrtl){
                 if($debug) Debug.write("reInit : resetting key");
             
                 this.setAttribute('key', '');
Index: WebContent/src/base/components/text/customEdittext.lzx
===================================================================
--- WebContent/src/base/components/text/customEdittext.lzx	(revision 4707)
+++ WebContent/src/base/components/text/customEdittext.lzx	(working copy)
@@ -2,7 +2,11 @@
 <library>
 
 <class name="customEdittext" extends="inputtext" bgcolor="0xFFFFFF" fontsize="11" height="20" >
-	
+
+    <include href="rtl.lzx" />
+    
+    <attribute name="isrtl" value="false" type="boolean"/>
+
 	<view name="v_border" height="1"       bgcolor="black" width="$once{parent.width}" />
     <view name="h_border" width="1"  y="1" bgcolor="black" height="$once{parent.height -1}" />
 	
@@ -18,6 +22,14 @@
 	<handler name="onfocus" args="f">
     	this.preText = this.getText();
     </handler>
+
+    <handler name="onisrtl">
+        if(this.isrtl){
+            this.setAttribute('textalign', 'right');
+        } else{
+            this.setAttribute('textalign', 'left');
+        }
+    </handler>
     
     <handler name="oninit">
     	//this.field.setAttribute("bgcolor",0xFFFFFF);
@@ -26,19 +38,23 @@
     <handler name="ontext" args="txt">
     	<![CDATA[
             
-    	    //
-    	   
+    	    //if ($debug) Debug.write("ontext", txt);
+
+    	    if (txt == null || txt.length == 0) {
+	        	return;
+	        }
+
+            if(this.isrtl) {
+	            this.setAttribute('text', this.rtl(txt));
+	        }
+
     	    if (System.capabilities.os.indexOf('Linux')==-1) { 
                 return; 
 	        }
 	        
 	        //if ($debug) Debug.write("System.capabilities.os -- START ",System.capabilities.os,txt);
 	        
-	        if (txt == null || txt.length == 0) {
-	        	return;
-	        }
 	        
-	        
 	        //UTF8 character remappings
             var cp1252 = new Object();
             cp1252[0x20ac]=0x80;
@@ -75,9 +91,10 @@
                     
             var i=Selection.getCaretIndex()-1; 
             
-            //if ($debug) Debug.write("i",i,Selection.getCaretIndex());
+            if ($debug) Debug.write("i",i,Selection.getCaretIndex());
             
             var d=i;
+
             
             //if ($debug) Debug.write("DO WHILE -1- ",t.charCodeAt(i),cp1252[t.charCodeAt(i)],i,0x80,0xBF);
             
@@ -122,7 +139,6 @@
             }
             
             //if ($debug) Debug.write("-- END ",this.text);
-            
     	]]>
     </handler>
 	
@@ -185,7 +201,6 @@
         </handler>
     </browserjavascriptcallback>    	
 
-    
     <method name="setLabelId" args="_labelid" >
         this.labelid = _labelid;
         this.setAttribute('text',canvas.getLabelName(this.labelid));
Index: WebContent/src/base/components/text/customInputtext.lzx
===================================================================
--- WebContent/src/base/components/text/customInputtext.lzx	(revision 4707)
+++ WebContent/src/base/components/text/customInputtext.lzx	(working copy)
@@ -2,39 +2,44 @@
 <library>
 
 <class name="customInputtext" extends="inputtext">
-	  
-	 <handler name="onfocus">
-	 
-	 	if ($debug) Debug.write("customInputtext onFocus : ", canvas.isrtl);
-	 		
-	 
-		 if(canvas.isrtl == 'true'){
-	 		if ($debug) Debug.write("customInputtext onFocus : RTL is set");
-	 		this.setAttribute('textalign', 'right');
+
+    <include href="rtl.lzx"/>
+    
+    <attribute name="isrtl" value="false" type="boolean"/>
+
+    <handler name="oninit">
+        if(this.isrtl){
+            this.setAttribute('textalign', 'right');
+        } else{
+            this.setAttribute('textalign', 'left');
         }
-        else{
-        	if ($debug) Debug.write("customInputtext onFocus : RTL is NOT set");
-      		this.setAttribute('textalign', 'left');
-        
+    </handler>
+
+    <handler name="onisrtl">
+        if(this.isrtl){
+            this.setAttribute('textalign', 'right');
+        } else{
+            this.setAttribute('textalign', 'left');
         }
-	 </handler>	  
-	  
-	 <handler name="ontext" args="txt">
+    </handler>
+
+    <handler name="ontext" args="txt">
         <![CDATA[
-            
-            //
-           
-            if (System.capabilities.os.indexOf('Linux')==-1) { 
-                return; 
-            }
-            
-            //if ($debug) Debug.write("System.capabilities.os -- START ",System.capabilities.os,txt);
-            
-            if (txt == null || txt.length == 0) {
+            //if ($debug) Debug.write("ontext", txt);
+
+    	    if (txt == null || txt.length == 0) {
+	        	return;
+	        }
+
+            if(this.isrtl) {
+	            this.setAttribute('text', this.rtl(txt));
+	        }
+
+    	    if (System.capabilities.os.indexOf('Linux')==-1) {
                 return;
-            }
-            
-            
+	        }
+
+
             //UTF8 character remappings
             var cp1252 = new Object();
             cp1252[0x20ac]=0x80;
@@ -64,35 +69,35 @@
             cp1252[0x0153]=0x9c;
             cp1252[0x017e]=0x9e;
             cp1252[0x0178]=0x9f;
-        
-            
-            var s=eval(Selection.getFocus()); 
+
+
+            var s=eval(Selection.getFocus());
             var t=txt;
-                    
-            var i=Selection.getCaretIndex()-1; 
-            
+
+            var i=Selection.getCaretIndex()-1;
+
             //if ($debug) Debug.write("i",i,Selection.getCaretIndex());
-            
+
             var d=i;
-            
+
             //if ($debug) Debug.write("DO WHILE -1- ",t.charCodeAt(i),cp1252[t.charCodeAt(i)],i,0x80,0xBF);
-            
+
             while (((t.charCodeAt(i)>=0x80 && t.charCodeAt(i)<=0xBF) || (cp1252[t.charCodeAt(i)])) && i>0) {
-                
+
                 //if ($debug) Debug.write("DO WHILE -2- ",t.charCodeAt(i),cp1252[t.charCodeAt(i)],i);
-                
+
                 if (cp1252[t.charCodeAt(i)]) {
                     t=t.substr(0,i)+String.fromCharCode(cp1252[t.charCodeAt(i)])+t.substr(i+1);
                 }
                 i--;
             }
-            if (i==d) { 
+            if (i==d) {
                 //if ($debug) Debug.write("i == d NOTHING TO REPLACE!");
-                return; 
+                return;
             }
-        
+
             var u=0;
-            
+
             if (t.charCodeAt(i)>=0xC2 && t.charCodeAt(i)<=0xDF && d-i==1) {
                  // two-byte sequence
                 u= (t.charCodeAt(i+1) & 0x3F)       +
@@ -106,23 +111,22 @@
                   ((t.charCodeAt(i+1) & 0x3C) << 6 ) +
                   ((t.charCodeAt(i  ) & 0x0F) << 12);
             }
-        
+
             if (u!=0) {
                 //if ($debug) Debug.write("Replace String -1- ",txt,u,"Replace With ",String.fromCharCode(u));
-                
+
                 this.setAttribute("text",t.slice(0,i)+String.fromCharCode(u)+t.slice(d+1));
-                
+
                 //if ($debug) Debug.write("Replace String -2- ",txt,u);
-                
+
                 this.setAttribute("text",this.text.split(String.fromCharCode(0x03)).join(''));
             }
-            
+
             //if ($debug) Debug.write("-- END ",this.text);
-            
+
         ]]>
     </handler>
-  	
-    
+
 </class>
 
 </library>
Index: WebContent/src/base/components/text/rtl.lzx
===================================================================
--- WebContent/src/base/components/text/rtl.lzx	(revision 0)
+++ WebContent/src/base/components/text/rtl.lzx	(revision 0)
@@ -0,0 +1,33 @@
+<method name="rtl" args="txt"><![CDATA[
+    var caret_i = Selection.getCaretIndex()-1;
+        //if ($debug) Debug.write("caret_i: ", caret_i, "txt.charAt(caret_i): ", txt.charAt(caret_i));
+        if(txt.charAt(caret_i) == ' ') {
+            if(caret_i == txt.length-1 || caret_i < txt.length-1 && txt.charAt(caret_i+1) == ' ') {
+                //if ($debug) Debug.write("txt.substr(0, caret_i): ", txt.substr(0, caret_i));
+                //if ($debug) Debug.write("(caret_i < txt.length-1): ", (caret_i < txt.length-1));
+                //if ($debug) Debug.write("txt.slice(caret_i + 1, txt.length): ", txt.slice(caret_i + 1, txt.length));
+                txt = txt.substr(0, caret_i) + ((caret_i < txt.length-1) ? txt.slice(caret_i + 1, txt.length) : "");
+                //if ($debug) Debug.write("remove whitespace: ", txt);
+            }
+            var found = false;
+            for(var i = caret_i-1; i>=0; i--) {
+                if(txt.charAt(i) == ' ') {
+                    found = true;
+                    txt = txt.slice(0, i) + " " + txt.slice(i + 1, txt.length)
+                    //if ($debug) Debug.write("next word: ", i, txt);
+                    Selection.setSelection(i,i);
+                    break;
+                } else if(txt.charAt(i) == '\n' || txt.charAt(i) == '\r') {
+                    found = true;
+                    txt = txt.slice(0, i+1) + " " + txt.slice(i + 1, txt.length)
+                    Selection.setSelection(i+1,i+1);
+                    break;
+                }
+            }
+            if(!found) {
+                txt = " " + txt;
+                Selection.setSelection(0,0);
+            }
+        }
+        return txt;
+]]></method>
Index: WebContent/src/base/hibernate/hibRtmpConnection.lzx
===================================================================
--- WebContent/src/base/hibernate/hibRtmpConnection.lzx	(revision 4707)
+++ WebContent/src/base/hibernate/hibRtmpConnection.lzx	(working copy)
@@ -234,6 +234,7 @@
 			userlang = canvas.language_id;
 		} else if (userlang == -1) {
 			var lang_id = canvas.default_lang_id;
+			var rtl = canvas.isrtl;
 			
 			var browserLang = flash.external.ExternalInterface.call("function(){return window.navigator.systemLanguage || window.navigator.userLanguage || window.navigator.language;}");
 			if (browserLang == null) {
@@ -266,11 +267,15 @@
 					i = (fullMatch != -1) ? fullMatch : partMatch;
 					xpath = "languagesSet:/item[" + i + "]/language_id/text()";
 					lang_id = languagesSet.getPointer().xpathQuery(xpath);
+					xpath = "boolean(languagesSet:/item[" + i + "]/rtl/text())";
+					rtl = languagesSet.getPointer().xpathQuery(xpath);
+					if($debug) Debug.write("RTL: ", rtl);
 				}
 			}
-			
             if (canvas.language_id == 0) {
+
             	userlang = lang_id;
+            	canvas.isrtl = rtl;
             } else {
             	userlang = canvas.language_id;
             }
@@ -488,6 +493,17 @@
             ]]>
         </handler>   
     </netRemoteCallHib>
+
+    <netRemoteCallHib name="getLanguage" funcname="languageservice.getLanguage" >
+        <netparam><method name="getValue">return parent.parent.userlang;</method></netparam>
+        <handler name="ondata" args="value">
+        <![CDATA[
+            if($debug) Debug.write("hibRtmpConnection/getLanguage: ", value);
+            canvas.setAttribute('isrtl', value.rtl);
+            parent.getLanguageByIdAndMax.doCall();
+        ]]>
+        </handler>
+    </netRemoteCallHib>
         
     <netRemoteCallHib name="getLanguageByIdAndMax" funcname="languageservice.getLanguageByIdAndMax" >
     	<attribute name="start" value="0" type="number" />
@@ -499,7 +515,7 @@
         <handler name="ondata" args="value">
         	<![CDATA[
 			if($debug) Debug.write("hibRtmpConnection/userlang:[",parent.userlang,"]");
-        		//if($debug) Debug.write("hibRtmpConnection/getLanguageByIdAndMax:[",value,"]");
+        		if($debug) Debug.write("hibRtmpConnection/getLanguageByIdAndMax:[",value,"]");
         		setLabelObjectByHundred(this.start,value);
 	            //The onResult-Handler will be called be the rtmpconnection
 	            if (this.start <= max){
@@ -784,7 +800,7 @@
                             	
                             }
                         } else {
-                        	hib.getLanguageByIdAndMax.doCall();
+                        	hib.getLanguage.doCall();
                         	canvas.focusOnItem = null;
                         }
                     } else {
Index: WebContent/src/base/auth/checkLoginData.lzx
===================================================================
--- WebContent/src/base/auth/checkLoginData.lzx	(revision 4707)
+++ WebContent/src/base/auth/checkLoginData.lzx	(working copy)
@@ -373,7 +373,7 @@
                             }
                         } else {
                         	
-                        	hib.getLanguageByIdAndMax.doCall();
+                        	hib.getLanguage.doCall();
                         	canvas.focusOnItem = null;
                             parent.close();
                             
Index: build.xml
===================================================================
--- build.xml	(revision 4707)
+++ build.xml	(working copy)
@@ -157,7 +157,7 @@
 		<mkdir dir="${main.out.dir}" />
 	</target>
 	
-	<condition property="db.provider" value="${db}" else="derby">
+	<condition property="db.provider" value="${db}" else="mysql">
 		<isset property="db" />
 	</condition>
 
@@ -460,7 +460,7 @@
 		</signjar>
 	</target>
 
-	<target name = "compile.laszlo.main" depends="compile.laszlo.main.debug">
+    <target name = "compile.laszlo.rtl">
 		<java classname = "org.openlaszlo.compiler.Main" fork = "true"
 	        newenvironment="true"
 	        failonerror="true"
@@ -472,6 +472,22 @@
 			<jvmarg value = "-Dlzproxied=${laszlo.proxied}"/>
 			<jvmarg value = "-Xms1024M"/>
 			<jvmarg value = "-Xmx1024M"/>
+			<arg line="--runtime=swf8 --dir ${out.dir.swf} --mcache on --onerror warn rtl.lzx" />
+		</java>
+	</target>
+
+	<target name = "compile.laszlo.main">
+		<java classname = "org.openlaszlo.compiler.Main" fork = "true"
+	        newenvironment="true"
+	        failonerror="true"
+	        dir = "${laszlo.src.dir}"
+	        classpathref="laszlo.lib"
+	       >
+			<jvmarg value = "-DLPS_HOME=${laszlo.home}"/>
+			<jvmarg value = "-Dlzdebug=${laszlo.debug}"/>
+			<jvmarg value = "-Dlzproxied=${laszlo.proxied}"/>
+			<jvmarg value = "-Xms1024M"/>
+			<jvmarg value = "-Xmx1024M"/>
 			<arg line="--runtime=swf8 --dir ${out.dir.swf} --mcache on --onerror warn main.lzx" />
 		</java>
 	</target>
